lib/fog/compute/kubevirt.rb

```diff
@@ -84,26 +137,162 @@ def initialize(options={})
           # Kubeclient needs different client objects for different API groups. We will keep in this hash the
           # client objects, indexed by API path/version.
           @clients = {}
+        end
+
+        def virt_supported?
+          virt_enabled = false
 
-          @client = kubevirt_client()
+          begin
+            api_versions = kubevirt_client.api["versions"]
+            api_versions.each do |ver|
+              if ver["groupVersion"]&.start_with?(KUBEVIRT_GROUP)
+                virt_enabled = true
+              end
+            end
+          rescue => err
+            # we failed to communicate or to evaluate the version format
+            @log.warn("Failed to detect kubevirt on provider with error: #{err.message}")
+          end
 
-          # TODO expect a specific token
-          @oc_client = openshift_client()
-          @kube_client = kube_client()
+          virt_enabled
         end
 
-        private
+        def valid?
+          kube_client.api_valid?
+        end
 
-        def client
-          @client
+        class WatchWrapper
+
+          def initialize(watch, mapper)
+            @watch = watch
+            @mapper = mapper
+          end
+
+          def each
+            @watch.each do |notice|
+              yield @mapper.call(notice)
+            end
+          end
+
+          def finish
+            @watch.finish
+          end
         end
 
-        def oc_client
-          @oc_client
+        #
+        # Returns a watcher for nodes.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_nodes(opts = {})
+          mapper = Proc.new do |notice|
+            if notice.object.kind == 'Node'
+              node = OpenStruct.new(Node.parse(notice.object))
+            else
+              node = OpenStruct.new
+            end
+            populate_notice_attributes(node, notice)
+            node
+          end
+          watch = kube_client.watch_nodes(opts)
+
+          WatchWrapper.new(watch, mapper)
         end
 
-        def kube_client
-          @kubeclient
+        #
+        # Returns a watcher for offline virtual machines.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_offline_vms(opts = {})
+          mapper = Proc.new do |notice|
+            if notice.object.kind == 'OfflineVirtualMachine'
```

Comment:
same as above, and maybe if we have many we should have something with method missing or some something that builds this procs (i am not familiar with the domain so not even trying with naming)
def build_proc(klass)
  Proc.new do |notice|
    obj = OpenStruct.new(Node.parse(notice.object)) if notice.object.kind == klass
    obj ||= OpenStruct.new
    populate_notice_attributes(obj, notice)
    obj
  end
end

Reply:
Will check.